pkg/control/addr.go

```diff
@@ -2,37 +2,47 @@
 // or more contributor license agreements. Licensed under the Elastic License;
 // you may not use this file except in compliance with the Elastic License.
 
-//go:build !windows
-
 package control
 
 import (
 	"crypto/sha256"
 	"fmt"
 	"path/filepath"
 
-	"github.com/elastic/elastic-agent/internal/pkg/agent/application/info"
 	"github.com/elastic/elastic-agent/internal/pkg/agent/application/paths"
-	"github.com/elastic/elastic-agent/pkg/utils"
 )
 
 // Address returns the address to connect to Elastic Agent daemon.
 func Address() string {
-	// when installed the control address is fixed
-	if info.RunningInstalled() {
-		root, _ := utils.HasRoot() // error is ignored
-		if root {
-			return paths.ControlSocketPath
-		}
-		return paths.ControlSocketUnprivilegedPath
-	}
+	return paths.ControlSocket()
+}
 
-	// unix socket path must be less than 104 characters
-	path := fmt.Sprintf("unix://%s.sock", filepath.Join(paths.TempDir(), "elastic-agent-control"))
-	if len(path) < 104 {
-		return path
+// AddressFromPath returns the connection address for an Elastic Agent running on the defined platform, and its
+// executing directory.
+func AddressFromPath(platform string, path string) (string, error) {
+	// elastic-agent will always derive the path to the socket using an absolute path
+	absDir, err := filepath.Abs(path)
+	if err != nil {
+		return "", fmt.Errorf("failed to get absolute path of %s: %w", path, err)
+	}
+	// elastic-agent is given a path from the OS that removes all symlinks, without then
+	// the path to the socket will not be the same.
+	noSyms, err := filepath.EvalSymlinks(absDir)
+	if err != nil {
+		return "", fmt.Errorf("failed to evaluate all symlinks of %s: %w", absDir, err)
+	}
+	// socket should be inside this directory
+	socketPath := filepath.Join(noSyms, paths.ControlSocketName)
+	if platform == "windows" {
+		// on windows the control socket is always at the sha256 of the socketPath
+		return fmt.Sprintf(`\\.\pipe\elastic-agent-%x`, sha256.Sum256([]byte(socketPath))), nil
+	}
+	unixSocket := fmt.Sprintf("unix://%s", socketPath)
+	if len(unixSocket) < 104 {
+		// small enough to fit
+		return unixSocket, nil
 	}
 	// place in global /tmp to ensure that its small enough to fit; current path is way to long
 	// for it to be used, but needs to be unique per Agent (in the case that multiple are running)
-	return fmt.Sprintf(`unix:///tmp/elastic-agent/%x.sock`, sha256.Sum256([]byte(path)))
+	return fmt.Sprintf(`unix:///tmp/elastic-agent/%x.sock`, sha256.Sum256([]byte(socketPath))), nil
```

Comment:
[Blocker]
It's replicating the logic from `initialControlSocketPath`. Shouldn't it be extracted to a function and both use this single function? With the logic duplicated it's at least risking a bug if someone changes one but not the other

Reply:
I extracted the function to a single location.